library/src/main/scala/scala/scalajs/runtime/RuntimeLong.scala

```diff
@@ -55,531 +121,558 @@ final class RuntimeLong(
-    else if (h != 0 && m == 0 && l == 0 && (h & (h - 1)) == 0)
-      (31 - Integer.numberOfLeadingZeros(h)) + BITS01
-    else
-      -1
-  }
-
-  private def setBit(bit: Int) =
-    if (bit < BITS)
-      new RuntimeLong(l | (1 << bit), m, h)
-    else if (bit < BITS01)
-      new RuntimeLong(l, m | (1 << (bit - BITS)), h)
-    else
-      new RuntimeLong(l, m, h | (1 << (bit - BITS01)))
-
-  private def divMod(y: RuntimeLong): scala.scalajs.js.Array[RuntimeLong] = {
-    // scalastyle:off return
-    import scala.scalajs.js
-    if (y.isZero) throw new ArithmeticException("/ by zero")
-    else if (x.isZero) js.Array(Zero, Zero)
-    else if (y.isMinValue) {
-      // MinValue / MinValue == 1, rem = 0
-      // otherwise == 0, rem x
-      if (x.isMinValue) js.Array(One, Zero)
-      else js.Array(Zero, x)
+  // Cache the instances for some "literals" used in this implementation
+  val Zero = new RuntimeLong(0, 0)
+  val One = new RuntimeLong(1, 0)
+  val MinusOne = new RuntimeLong(-1, -1)
+  val MinValue = new RuntimeLong(0, 0x80000000)
+  val MaxValue = new RuntimeLong(0xffffffff, 0x7fffffff)
+
+  def fromDouble(value: Double): RuntimeLong = {
+    import Utils._
+
+    if (value.isNaN) {
+      Zero
+    } else if (value < -TowPow63) {
+      MinValue
+    } else if (value >= TowPow63) {
+      MaxValue
     } else {
-      val xNegative = x.isNegative
-      val yNegative = y.isNegative
-
-      val xMinValue = x.isMinValue
-
-      val absY = y.abs
-
-      val pow = absY.powerOfTwo
-      if (pow >= 0) {
-        if (xMinValue) {
-          val z = x >> pow
-          js.Array(if (yNegative) -z else z, Zero)
-        } else {
-          // x is not min value, so we can calculate absX
-          val absX = x.abs
-          val absZ = absX >> pow
-          val z = if (xNegative ^ yNegative) -absZ else absZ
-          val remAbs = absX.maskRight(pow)
-          val rem = if (xNegative) -remAbs else remAbs
-          js.Array(z, rem)
-        }
-      } else {
-        val newX = {
-          if (xMinValue) {
-            /* In theory we would need 2^31 == MaxValue + 1 here, but that
-             * overflows. Instead, we use a trick.
-             * Since absY is not 2^k for any k at this point, 2^31 cannot
-             * possibly be exactly divided by k, so
-             * 2^31 / absY == (2^31 - 1) / absY == MaxValue / absY
-             * The remainder will be off by 1, but divModHelper fixes that
-             * after the fact.
-             */
-            MaxValue
-          } else {
-            val absX = x.abs
-            if (absX < absY)
-              return js.Array(Zero, x) // <-- ugly but fast
-            else
-              absX
-          }
-        }
-        divModHelper(newX, absY, xNegative, yNegative, xMinValue)
-      }
+      val neg = value < 0
+      val absValue = if (neg) -value else value
+      val lo = rawToInt(absValue)
+      val hi = rawToInt(absValue / TwoPow32)
```

Comment:
Actually, it seems that the `rawToInt(lo)` / `rawToInt(hi / TwoPow32)` is also a candidate for an inlined helper.